api/api/models_api.go

```diff
@@ -99,3 +104,126 @@ func (c *ModelsController) GetModel(r *http.Request, vars map[string]string, bod
 
 	return Ok(model)
 }
+
+// DeleteModel delete a model given a project and model ID
+func (c *ModelsController) DeleteModel(r *http.Request, vars map[string]string, body interface{}) *Response {
+	ctx := r.Context()
+
+	projectID, err := models.ParseID(vars["project_id"])
+	if err != nil {
+		return BadRequest("Unable to parse project_id")
+	}
+	modelID, err := models.ParseID(vars["model_id"])
+	if err != nil {
+		return BadRequest("Unable to parse model_id")
+	}
+
+	_, err = c.ProjectsService.GetByID(ctx, int32(projectID))
+	if err != nil {
+		return NotFound(err.Error())
+	}
+
+	model, err := c.ModelsService.FindByID(ctx, modelID)
+	if err != nil {
+		if gorm.IsRecordNotFoundError(err) {
+			return NotFound(fmt.Sprintf("Model is not found: %s", err.Error()))
+		}
+		return InternalServerError(err.Error())
+	}
+
+	// get all model version for this model
+	versions, _, err := c.VersionsService.ListVersions(ctx, modelID, c.MonitoringConfig, service.VersionQuery{})
+	if err != nil {
+		return InternalServerError(err.Error())
+	}
+
+	var allJobInModel [][]*models.PredictionJob
+	var allEndpointInModel [][]*models.VersionEndpoint
+
+	// iterate for every version, check the active prediction job / active endpoint
+	for _, version := range versions {
+		if model.Type == "pyfunc_v2" {
+			// check active prediction job
+			// if there are any active prediction job using the model version, deletion of the model version are prohibited
+			jobs, response := c.VersionsController.getInactivePredictionJobsForDeletion(ctx, model, version)
+			if response != nil {
+				return response
+			}
+
+			// save all prediction jobs for deletion process later on
+			allJobInModel = append(allJobInModel, jobs)
+
+		} else {
+			// handle for model with type non pyfunc
+			// check active endpoints
+			// if there are any active endpoint using the model version, deletion of the model version are prohibited
+			endpoints, response := c.VersionsController.getInactiveEndpointsForDeletion(ctx, model, version)
+			if response != nil {
+				return response
+			}
+
+			// save all endpoint for deletion process later on
+			allEndpointInModel = append(allEndpointInModel, endpoints)
+
+		}
+	}
+
+	// DELETING ALL THE RELATED ENTITY
+	for index, version := range versions {
+		if model.Type == "pyfunc_v2" {
+			// DELETE PREDICTION JOBS
+			response := c.VersionsController.deleteInactivePredictionJobs(ctx, allJobInModel[index], model, version)
```

Comment:
CMIIW, but @zenovore, this `[index]` is only correct if each version has exactly 1 job and 1 endpoint. Eg: If there are 0 endpoints for a version or 2 (because the same version can be deployed in multiple environments), this logic fails.

Reply:
Ok scratch that comment. I just realised that the `allJobInModel` and `allEndpointInModel` are array of array. So, this is fine.